src/Confluent.Kafka/Consumer.cs

```diff
@@ -129,40 +131,65 @@ private void LogCallback(IntPtr rk, SyslogLevel level, string fac, string buf)
             {
-                if (rebalanceHandler != null)
+                if (partitionsAssignedHandler == null)
+                {
+                    Assign(partitionAssignment.Select(p => new TopicPartitionOffset(p, Offset.Unset)));
+                    return;
+                }
+
+                lock (assignCallCountLockObj) { assignCallCount = 0; }
+                var assignTo = partitionsAssignedHandler(partitionAssignment);
+                lock (assignCallCountLockObj)
                 {
-                    lock (assignCallCountLockObj) { assignCallCount = 0; }
-                    rebalanceHandler(new RebalanceEvent(partitionAssignment, true));
-                    lock (assignCallCountLockObj)
+                    if (assignCallCount > 0)
                     {
-                        if (assignCallCount > 0) { return; }
+                        throw new InvalidOperationException("Assign/Unassign must not be called in the partitions assigned handler.");
                     }
                 }
-                Assign(partitionAssignment.Select(p => new TopicPartitionOffset(p, Offset.Invalid)));
+                Assign(assignTo);
+                return;
             }
-            else if (err == ErrorCode.Local_RevokePartitions)
+            
+            if (err == ErrorCode.Local_RevokePartitions)
             {
-                if (rebalanceHandler != null)
+                if (partitionsRevokedHandler == null)
+                {
+                    Unassign();
+                    return;
+                }
+
+                lock (assignCallCountLockObj) { assignCallCount = 0; }
+                var assignmentWithPositions = new List<TopicPartitionOffset>();
+                foreach (var tp in partitionAssignment)
                 {
-                    lock (assignCallCountLockObj) { assignCallCount = 0; }
-                    rebalanceHandler(new RebalanceEvent(partitionAssignment, false));
-                    lock (assignCallCountLockObj)
+                    try
+                    {
+                        assignmentWithPositions.Add(new TopicPartitionOffset(tp, Position(tp)));
+                    }
+                    catch
                     {
-                        if (assignCallCount > 0) { return; }
+                        assignmentWithPositions.Add(new TopicPartitionOffset(tp, Offset.Unset));
                     }
                 }
-                Unassign();
-            }
-            else
-            {
-                throw new KafkaException(kafkaHandle.CreatePossiblyFatalError(err, null));
+                var assignTo = partitionsRevokedHandler(assignmentWithPositions);
+                lock (assignCallCountLockObj)
+                {
+                    if (assignCallCount > 0)
+                    {
+                        throw new InvalidOperationException("Assign/Unassign must not be called in the partitions revoked handler.");
```

Comment:
What about reversing the logic here and throwing the exception from assign instead?
And the current locking does not really work for the case where assign() is called from another thread while the rebalance handler is executing.
You could make assignCallCount a TLS variable instead to see if it was called from the handler or not.

Reply:
yep, all good points. it's like this because I just morphed earlier code. in the interests of velocity, i want to leave this as is (and will open an issue to track) - it's just a check for application logic errors and it'll work pretty much always.